### PR TITLE
cpu/stm32: Fix garbage on UART init

### LINKED ISSUE #8045: stm32/periph/uart: extra byte transmitted on first transmission
After testing the slipdev driver on a nucleo-f446, it appeared to me as if an extra 0xFF byte is transmitted by the UART driver. As somewhat described in https://github.com/RIOT-OS/RIOT/pull/8041#issuecomment-344651841, this can be explicitly triggered with slipdev by having the node transmit a frame. An initial 0xff byte should appear as first byte (and by doing so crash the tunslip6 application).

More a suspicion than a confirmed bug, it would help if somebody can reproduce this (or refute this bug), I still have to test if this 0xff byte is also added to the UART interface of the shell.

Oh and close and ignore this issue if it is caused by the low on reset nature of the hardware and the active-low of UART.

### PR DESCRIPTION
**Pull request by maribu**
#### cpu/stm32: Fix gpio_init() / gpio_int_af()
    
- Do not set an intermediate mode, prepare correct mode settings in a temporary variable
- Consistently enabled the GPIO periph in gpio_init_af()
    - Previously, STM32 F1 did not require a separate call to gpio_init() prior to a call of gpio_init_af(), but other STM32 families did
    - Now, gpio_init_af() can be used without gpio_init() consistently
- STM32 F1: Do not touch ODR for non input pins
    - For input pins, this enables / disabled pull up resistors. For outputs, this register should remain untouched (according to API doc)

#### cpu/stm32: Fix uart_init()
    
- Make use of the fact that gpio_init_af() does not need prior call to gpio_init() for all STM32 families anymore and drop call to gpio_init()
- Initialize the UART periph first, before initializing the pins
    - While uninitialized, the UART periph will send signal LOW to TXD. This results in a start bit being picked up by the other side.
    - Instead, we do not connect the UART periph to the pins until it is initialized, so that the TXD level will already be HIGH when the pins are attached.
    - This results in no more garbage being send during initialization

### Testing procedure

- UART should still work for STM32F1 and STM32Fx (x != 1)
- `gpio_init()` should still work for ST32F1 and STM32Fx (x != 1)

Note: It should be sufficient to test with STM32F1 and one of the other STM32 not belonging to the STM32F1 family

### Issues/PRs references

Fixes #8045

### CODE REVIEW

File: cpu/stm32/periph/gpio_all.c [hunk 1]
```diff
@@ -122,6 +126,16 @@ int gpio_init(gpio_t pin, gpio_mode_t mode)
     return 0;
 }
 
+int gpio_init(gpio_t pin, gpio_mode_t mode)
+{
+    return _gpio_init(pin, mode, 0);
+}
+
+int gpio_init_set(gpio_t pin, gpio_mode_t mode)
```
Comment (benpicco) [inline cpu/stm32/periph/gpio_all.c:134]: How about `gpio_init_high()` :wink:

**Comment by fjmolinas** [comment]
Is `�` in:

```
2020-07-03 17:07:25,205 # �main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-pr-14426)
2020-07-03 17:07:25,206 # Hello World!
2020-07-03 17:07:25,211 # You are running RIOT on a(n) nucleo-f103rb board.
2020-07-03 17:07:25,214 # This board features a(n) stm32 MCU.
```

the garbage? do I need to interface directly over uart or though the debugger works? Or do I need a logic analyzer?

**Comment by maribu** [comment]
> Is `�` in:

Yep, exactly. With a logic analyzer you should be able to confirm that the initialization is only "almost flicker free", as the pin sadly still goes briefly down during initialization. However, with this PR the duration of the pin being LOW during initialization is much shorter. On my Nucleo F446RE, I tested booting about 50 times and the UART interface of the debugger never picked up this as a start bit, while it previously got a 0xff character on every boot.

For testing this PR just using `make term` should be sufficient. The garbage character before `main()` in the boot message should no longer appear with this PR.

**Comment by fjmolinas** [comment]
> For testing this PR just using make term should be sufficient. The garbage character before main() in the boot message should no longer appear with this PR.

In that case just with make term I still get the garbage `�` in `stm32f1`, specifically `nucleo-d103rb` when using `pyterm` when just having `make term` in a terminal and flashing from another. I think make term is not reliable for this.

**Comment by maribu** [comment]
> In that case just with make term I still get the garbage `�` in `stm32f1`

It seems I god lucky with my UART adapter:

<details><summary>Console output (multiple reboots triggered via reset button)</summary>

```
2020-07-03 22:38:15,881 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:15,882 # Hello World!
2020-07-03 22:38:15,887 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:15,890 # This board features a(n) stm32 MCU.
2020-07-03 22:38:16,285 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:16,286 # Hello World!
2020-07-03 22:38:16,291 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:16,294 # This board features a(n) stm32 MCU.
2020-07-03 22:38:25,596 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:25,597 # Hello World!
2020-07-03 22:38:25,602 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:25,605 # This board features a(n) stm32 MCU.
2020-07-03 22:38:26,195 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:26,196 # Hello World!
2020-07-03 22:38:26,201 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:26,204 # This board features a(n) stm32 MCU.
2020-07-03 22:38:26,721 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:26,722 # Hello World!
2020-07-03 22:38:26,727 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:26,730 # This board features a(n) stm32 MCU.
2020-07-03 22:38:27,149 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:38:27,151 # Hello World!
2020-07-03 22:38:27,155 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:38:27,159 # This board features a(n) stm32 MCU.
2020-07-03 22:39:41,864 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:39:41,865 # Hello World!
2020-07-03 22:39:41,870 # You are running RIOT on a(n) bluepill-128kib board.
2020-07-03 22:39:41,873 # This board features a(n) stm32 MCU.
2020-07-03 22:40:04,588 # main(): This is RIOT! (Version: 2020.07-devel-1726-gadffa6-stm32f4_uart_init)
2020-07-03 22:40:04,590 # Welcome to RIOT!
```

</details>

With the logic analyzer I can however see a start bit long enough that it should actually be detected by my UART adapter.

**Comment by maribu** [comment]
OK. For the STM32F1xx I found the culprits:

1. The UART needs to be initialized first, and only afterwards the UART needs to be connected to the pins.
2. The ODR register is both used for output in case of pins configured as output, and for pull up in case of input pins. STM32F1 `gpio_init()` will clear the ODR register for all modes but `GPIO_IN_PU`, it should however not touch the ODR register for non-input modes.

I think I don't need `gpio_init_high()` to fix the UART issue anymore, as the pin should remain high when the UART is initialized prior to routing it to the output pins. Still, other use cases require that PR.

I will update this PR when I have time to clean up the WIP code.

**Comment by maribu** [comment]
I updated the fix and force pushed, as this is basically a rewrite compared to the original approach. Now the result is correct even when using a logic analyzer. There should be no garbage being detected anymore.

I think this fix might need a bit more testing, as a broken GPIO implementation might result in very unhappy users.

Btw: I realized that `gpio_init()` is not thread safe. If one thread interrupts another during `gpio_init()` and performs itself `gpio_init()` on a pin of the same PORT, some of the configuration changes done by the higher prio thread might be reverted by the lower prio thread. (Lost update in a non-atomic read modify write sequence.)

I think the odds of this resulting in an issue are pretty low, as most GPIO initialization is done during `auto_init()` in sequential fashion. Two threads using bit-banging with half-duplex protocols are the only obvious example coming to my mind thread safety in `gpio_init()` would be needed. (E.g. two DHT sensors being read out concurrently.)

**Comment by maribu** [comment]
I'll rebase to fix a typo in a commit message. I wont change the commits or squash. (I'll also disable ready for build now: Now need to rebuild for a changed commit message.)

**Comment by maribu** [comment]
Out of the bugs still open for the release, this is likely the least important one. The UART is working reliable after the first broken character without this. So we might focus on the other open PRs first.

(IMO, we might even just leave this out of the release. UART is known to not be super reliable anyway, so one if one byte of garbage more breaks some application, that application is desperately in need of some error detection mechanism anyway.)

**Comment by fjmolinas** [comment]
This now realiably fixes the issue for me on `stm32*`:

**PR**

- stm32f1

```
2020-07-15 11:38:17,008 # main(): This is RIOT! (Version: 2020.07-devel-1728-gb7f7-pr-14426)
2020-07-15 11:38:17,008 # Hello World!
2020-07-15 11:38:17,008 # You are running RIOT on a(n) iotlab-m3 board.
2020-07-15 11:38:17,023 # This board features a(n) stm32 MCU.
2020-07-15 11:38:21,184 # main(): This is RIOT! (Version: 2020.07-devel-1728-gb7f7-pr-14426)
2020-07-15 11:38:21,184 # Hello World!
2020-07-15 11:38:21,184 # You are running RIOT on a(n) iotlab-m3 board.
2020-07-15 11:38:21,185 # This board features a(n) stm32 MCU.
```

- stm32*

```
2020-07-15 11:45:19,381 # main(): This is RIOT! (Version: 2020.07-devel-1728-gb7f7-pr-14426)
2020-07-15 11:45:19,382 # Hello World!
2020-07-15 11:45:19,387 # You are running RIOT on a(n) nucleo-l152re board.
2020-07-15 11:45:19,388 # This board features a(n) stm32 MCU.
2020-07-15 11:45:25,118 # main(): This is RIOT! (Version: 2020.07-devel-1728-gb7f7-pr-14426)
2020-07-15 11:45:25,119 # Hello World!
2020-07-15 11:45:25,121 # You are running RIOT on a(n) nucleo-l152re board.
2020-07-15 11:45:25,123 # This board features a(n) stm32 MCU.
```

**master**

- stm32f1

```
2020-07-15 11:38:04,993 # �main(): This is RIOT! (Version: 2020.07-devel-1690-g772b6-HEAD)
2020-07-15 11:38:04,993 # Hello World!
2020-07-15 11:38:04,993 # You are running RIOT on a(n) iotlab-m3 board.
2020-07-15 11:38:04,994 # This board features a(n) stm32 MCU.
2020-07-15 11:38:08,208 # �main(): This is RIOT! (Version: 2020.07-devel-1690-g772b6-HEAD)
2020-07-15 11:38:08,209 # Hello World!
2020-07-15 11:38:08,224 # You are running RIOT on a(n) iotlab-m3 board.
2020-07-15 11:38:08,224 # This board features a(n) stm32 MCU.
```
- stm32*

```
2020-07-15 11:44:42,723 # �main(): This is RIOT! (Version: 2020.07-devel-1690-g772b6-HEAD)
2020-07-15 11:44:42,723 # Hello World!
2020-07-15 11:44:42,728 # You are running RIOT on a(n) nucleo-l152re board.
2020-07-15 11:44:42,730 # This board features a(n) stm32 MCU.
2020-07-15 11:44:47,468 # �main(): This is RIOT! (Version: 2020.07-devel-1690-g772b6-HEAD)
2020-07-15 11:44:47,469 # Hello World!
2020-07-15 11:44:47,475 # You are running RIOT on a(n) nucleo-l152re board.
2020-07-15 11:44:47,477 # This board features a(n) stm32 MCU.

```

**Comment by fjmolinas** [comment]
If uart works then `gpio_init` also works IMO.

**Comment by fjmolinas** [comment]
Please squash @maribu and trigger ci!

**Comment by aabadie** [comment]
@miri64 since this PR is fixing a bug, I think it needs a backport, right ?

**Comment by miri64** [comment]
Yepp.

**Pull request merged**